Fix: Make link unclickable. Closes #91

shumail92 opened the pull request:
Closes #91

marhi commented:
Not a very long term solution.

irmana commented:
@marhi if it's not a long term, why do we merge then? I suggest we discuss this internally and see what do we want to achieve here.

shumail92 commented:
@marhi Can you suggest a better approach ?

I am working on writing the test at the moment.

irmana commented:
@shumail92 we discussed this internally and I would propose the solution with configuration file where user can define which key combinations along with the mouse click make links clickable. How about that? So, by default links are not clickable, but if you press i.e. Ctrl + click then it opens up the link in a browser.

shumail92 commented:
@irmana Okay, I try for that. 
btw, I just updated my PR with test case for current behavior (un-clickable links).

irmana commented:
@shumail92 please submit a solution in 2 separate commits... 1st commit with the fix, 2nd commit with the test/s

shumail92 commented:
@irmana Sure. When I submit solution for new proposed approach, I'll split into 2 commits. Thanks

irmana commented:
@shumail92 please set the label in progress

shumail92 commented:
@irmana I don't have permissions to set labels.

marhi commented:
What's the progress? I have the solution for opening the link in OS browser, it's just three lines?
Maybe a few lines to make it a config file option.

jkuri commented:
i personally would prefer opening url in the browser.

-- 
Jan Kuri

On 25 May 2017 at 18:19:06, Tomaž Marhat (notifications@github.com) wrote:

> What's the progress? I have the solution for opening the link in OS
> browser, it's just three lines?
> Maybe a few lines to make it a config file option.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bleenco/bterm/pull/92#issuecomment-304053093>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABtntt_JWCLilcXNGRGLuMBFJptsh0xgks5r9an6gaJpZM4Nk9u_>
> .
>

omerson commented:
I would add none of the well known terminals support this.

jkuri commented:
@omerson whats your point? we are not copying iterm2 or something.

marhi commented:
I agree with that, but isn't it our goal here to be innovative and offer something more than others?

irmana commented:
What I would personally find intriguing is:
- By default if I click once it copies the link 
- When I press Ctrl and click then it opens the link in a browser

That would be fantastic I think.

omerson commented:
@marhi I agree just refering to standards and user expectations.

shumail92 commented:
I added the behavior that @irmana suggested. For the tests, I added a basic test for asserting that link doesn't open on clicking a url(without pressing ctrl or cmd on mac) in bterm window. 
I will also try to add a check for asserting if link is copied.

jkuri commented on src/app/components/window-terminal/window-terminal.component.ts:
```diff
@@ -50,6 +50,19 @@ export class WindowTerminalComponent implements OnInit {
       }, 800);
     }, false);
+
+    document.addEventListener('click', function (e) {
```
use `@HostListener`.

jkuri commented on src/app/components/window-terminal/window-terminal.component.ts:
```diff
@@ -50,6 +50,19 @@ export class WindowTerminalComponent implements OnInit {
+      if (e.target['tagName'] === 'A') {
+        e.preventDefault();
+        clipboard.writeText(e.target['href']);
+        let isMac = process.platform === 'darwin'
```
missing `;`

shumail92 commented:
I'm selecting the link using `this.app.client.click('.xterm-rows div:first-child a'))` and it's passing test on my macOS. But tests are failing on circleci/appveyor. 
Any idea please?

marhi commented:
Yes, your tests are very fragile. The presumtion that the link will be in the first line is rarely met. Usually there is stuff like MOTD printed / Windows version, etc in the first few lines...

Try with:
      .then(() => this.app.client.pause(2000))
      .then(() => this.app.client.click('.terminal-instance a'))

shumail92 commented:
Thankyou. I assume that because i clear terminal before writing link. In
case, there's some existing text (windows version etc) shouldn't it be
cleared with the `clear`?


On Fri, 26 May 2017 at 12:32, Tomaž Marhat <notifications@github.com> wrote:

> Yes, your tests are very fragile. The presumtion that the link will be in
> the first line is rarely met. Usually there is stuff like MOTD printed /
> Windows version, etc in the first few lines...
>
> Try with:
> .then(() => this.app.client.pause(2000))
> .then(() => this.app.client.click('.terminal-instance a'))
>
> —
> You are receiving this because you were assigned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/bleenco/bterm/pull/92#issuecomment-304248607>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AE3pinZiIZu-oVF9ZlQFx8illHnKggE2ks5r9qozgaJpZM4Nk9u_>
> .
>

jkuri commented on src/app/components/window-terminal/window-terminal.component.ts:
```diff
@@ -52,6 +52,19 @@ export class WindowTerminalComponent implements OnInit {
     }, false);
   }
 
+  @HostListener('click', ['$event']) handleURLClick(event) {
```
`(event: MouseEvent)` please be more specific.

shumail92 replied:
@jkuri Thank you for the review. I initially used `event: Mouseevent` but to avoid `Property tagName doesn't exist on type EventTarget` i changed it to Event. 
Is it fine if i leave those warnings like this? 
Or another workaround I was thinking to cast it inline like `(event.target as any).href` or `(event.target as any).tagName`. 
What would yo suggest

jkuri replied:
did you try `handleURLClick(event: MouseEvent)`?

shumail92 replied:
Yes, same warning.

jkuri replied:
okay, then just use `(event: any)` I think, it will be okay

shumail92 replied:
Sure, thanks. Updating.

jkuri commented on src/app/components/window-terminal/window-terminal.component.ts:
```diff
@@ -52,6 +52,19 @@ export class WindowTerminalComponent implements OnInit {
   }
 
+  @HostListener('click', ['$event']) handleURLClick(event) {
+    if (event.target['tagName'] === 'A') {
```
please use dot notation.

jkuri commented:
can you rebase your PR please?

shumail92 commented:
@jkuri Thanks. Sure I rebase it, on it.

jkuri commented:
thanks!
